core/local/steps/incomplete_fixer.js

```diff
@@ -0,0 +1,95 @@
+/* @flow */
+
+const path = require('path')
+
+const { id } = require('../../metadata')
+const stater = require('../stater')
+
+// Drop incomplete events after this delay (in milliseconds).
+// TODO tweak the value (the initial value was chosen because it looks like a
+//      good value, it is not something that was computed)
+const DELAY = 3000
+
+// TODO add unit tests and logs
+
+/*::
+import type Buffer from './buffer'
+import type { AtomWatcherEvent } from './event'
+import type { Checksumer } from '../checksumer'
+
+type IncompleteItem = {
+  event: AtomWatcherEvent,
+  timestamp: number,
+}
+*/
+
+// When a file is added or updated, and it is moved just after, the first event
+// is marked as incomplete by addChecksum because we cannot compute the
+// checksum at the given path. But the event is still relevant, in particular if
+// a directory that is an ancestor of this file has been moved. With the renamed
+// event, by comparing the path, we can extrapolate the new path and check with
+// fs.stats if we have a file here.
+//
+// Cf test/property/local_watcher/swedish_krona.json
+module.exports = function (buffer /*: Buffer */, opts /*: { syncPath: string , checksumer: Checksumer } */) /*: Buffer */ {
+  const incompletes = []
+  return buffer.asyncMap(async (events) => {
+    // Filter out the incomplete events
+    const batch = []
+    for (const event of events) {
+      if (event.incomplete) {
+        const now = new Date()
+        incompletes.push({ event, timestamp: +now })
+        continue
+      }
+      batch.push(event)
+    }
+
+    // Let's see if we can match an incomplete event with this renamed event
+    for (const event of batch) {
+      if (incompletes.length === 0 || event.action !== 'renamed') {
+        continue
+      }
+
+      const limit = +new Date() - DELAY
+      for (let i = 0; i < incompletes.length; i++) {
+        const item = incompletes[i]
+
+        // Remove the expired incomplete events
+        if (i === 0 && item.timestamp < limit) {
```

Comment:
The expired incomplete events are always at the start of the array. The condition is probably useless now. When I put it, the limit was not a variable but something rebuilt on each iteration, and I fear that if the loop was too slow, the limit was no longer the same for all the events, and that will make us remove the wrong events. Since then, I've moved the limit in a variable initialized outside of the loop, so, we can remove this condition (or keep it): it shouldn't matter.

Reply:
I'd rather remove it if it's unnecessary. Better avoid raising questions about it.

Reply:
Well, I have the opposite reaction: the code raises question in my mind when we do not check if `i` is 0 as we are doing a `shift` to remove an item. But, if you prefer without it, let's go for that.

Reply:
Ah I see what you mean. I didn't think about that at all :D. In this case, maybe we could replace `incomples.shift()` with `incompletes.splice(i, 1)`. This would probably get optimized by the browser.